examples/reporting/get_campaign_stats_to_csv.py

```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/env python
+#
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This examples gets a basic campaign stats report for a given customer ID
+and writes them to a CSV file.
+
+The LoadFromStorage method is pulling credentials and properties from a
+"googleads.yaml" file. By default, it looks for this file in your home
+directory. For more information, see the "Caching authentication information"
+section of our README.
+
+Examples:
+    Write to file output.csv in the same directory as script with headers.
+        $ python get_campaign_stats_to_csv.py -c 0123456789 -o output.csv -w
+
+    Write to file output.csv in the same directory as script without headers.
+        $ python get_campaign_stats_to_csv.py -c 0123456789 -o output.csv -w
+"""
+
+import argparse
+import csv
+import sys
+
+from google.ads.google_ads.client import GoogleAdsClient
+from google.ads.google_ads.errors import GoogleAdsException
+
+
+def main(client, customer_id, output_file, write_headers):
+    """Demonstrates how to write rows returned from a Google Ads search_stream
+    request to a CSV file.
+    Args:
+        client: An initialized GoogleAdsClient instance.
+        customer_id (str): The client customer ID string.
+        output_file (str): Filename of the file to write the report data to.
+        write_headers (bool): From argparse, True if provided, False otherwise.
+    """
+    ga_service = client.get_service("GoogleAdsService", version="v6")
+
+    query = """
+        SELECT
+          customer.descriptive_name,
+          segments.date,
+          campaign.name,
+          metrics.impressions,
+          metrics.clicks,
+          metrics.cost_micros
+        FROM campaign
+        WHERE
+          segments.date DURING LAST_7_DAYS
+        ORDER BY metrics.impressions DESC
+        LIMIT 25
+        """
+
+    # Issues a search request using streaming.
+    response = ga_service.search_stream(customer_id, query)
+    try:
+        with open(output_file, "w", newline="") as f:
+            writer = csv.writer(f)
+
+            # Define a list of headers for the first row.
+            headers = ["Account", "Date", "Campaign", "Impressions", "Clicks",
+                       "Cost"]
+
+            # If the write_headers flag was passed, write header row to the CSV
+            if write_headers:
+                writer.writerow(headers)
```

Comment:
Should this be `writer.writeheader`?

Reply:
I believe it's correct here. To use writeheader(), the writer must be initialized as a dictWriter(). This then means that the non header rows will have to be written as a dict instead of a list too. For a large enough response, the overhead is noticeable. I can convert the standard row writing to a map but I feel this is adding additional complexity.